src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -1099,6 +1103,25 @@ private void EmitArrayElementLoad(BoundArrayAccess arrayAccess, bool used)
             EmitPopIfUnused(used);
         }
 
+        private void EmitArrayElementRefLoad(BoundRefArrayAccess refArrayAccess, bool used)
+        {
+            if (used)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+
+            // Value types do not need runtime element type checks.
+            if (refArrayAccess.ArrayAccess.Type.IsValueType)
```

Comment:
>refArrayAccess.ArrayAccess.Type.IsValueType [](http://example.com/codeflow?start=16&length=43)

Why not put this check to Optimizer and avoid special node creation?

Reply:
Seemed like emit-layer concern about which instruction to emit, but it makes sense to avoid special node creation.

Reply:
We are checking `readonly ref` case in Optimizer, but this is also about an instruction choice